misk/src/main/kotlin/misk/hibernate/RealTransacter.kt

```diff
@@ -1,24 +1,44 @@
 package misk.hibernate
 
 import com.google.common.collect.ImmutableSet
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
 import misk.jdbc.DataSourceConfig
 import misk.jdbc.DataSourceType
 import org.hibernate.SessionFactory
+import org.hibernate.StaleObjectStateException
+import org.hibernate.exception.LockAcquisitionException
 import java.sql.Connection
 import java.sql.ResultSet
+import java.time.Duration
+import javax.persistence.OptimisticLockException
 import kotlin.reflect.KClass
 
-internal class RealTransacter(
+internal class RealTransacter private constructor(
   private val sessionFactory: SessionFactory,
-  private val config: DataSourceConfig
+  private val config: DataSourceConfig,
+  private val threadLocalSession: ThreadLocal<Session>,
+  private val options: TransacterOptions
 ) : Transacter {
 
-  private val threadLocalSession = ThreadLocal<Session>()
+  constructor(sessionFactory: SessionFactory, config: DataSourceConfig) :
+      this(sessionFactory, config, ThreadLocal(), TransacterOptions())
 
   override val inTransaction: Boolean
     get() = threadLocalSession.get() != null
 
   override fun <T> transaction(lambda: (session: Session) -> T): T {
+    return when {
+      options.maxAttempts > 1 -> {
+        val backoff = ExponentialBackoff(Duration.ZERO, Duration.ZERO)
+        retry(options.maxAttempts, backoff, ::isRetryable) { transactionInternal(lambda) }
+      }
+      options.maxAttempts == 1 -> transactionInternal(lambda)
+      else -> throw IllegalArgumentException()
+    }
```

Comment:
I find this really tough to follow

Reply:
can we make it so the retry function handles `maxAttempts == 1` by not retrying?

Reply:
fixed.